main.tf

```diff
@@ -8,11 +8,16 @@ locals {
 
   create_security_group    = local.enabled && length(var.target_security_group_id) == 0
   sg_create_before_destroy = var.create_before_destroy
+  # If the security group is not being created by this module, we need to treat it as
+  # needing to be preserved, because we cannot replace it here.
+  preserve_security_group_id = var.preserve_security_group_id && local.create_security_group
```

Comment:
Good catch, the logic is wrong. Partly because I hate the way Terraform formats the `not` operator, and partly because of how I handle the `enabled` flag.
```suggestion
  preserve_security_group_id = var.preserve_security_group_id || length(var.target_security_group_id) > 0
```